akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,377 @@
+
+      Behaviors.stopped
+    }
+
+  }
+
+  // the active phase switches between PersistingEvents and HandlingCommands;
+  // we do this via a var instead of behaviours to keep allocations down as this will be flip/flaping on every Persist effect
+  private[this] var phase: EventsourcedRunningPhase = HandlingCommands
+
+  override def onMessage(msg: Any): Behavior[Any] = {
+    log.info("RUNNING onMessage: " + msg + s" @ ${phase.name}")
+    msg match {
+      // TODO explore crazy hashcode hack to make this match quicker...?
+      case SnapshotterResponse(r) ⇒ onSnapshotterResponse(r)
+      case JournalResponse(r)     ⇒ phase.onJournalResponse(r)
+      case command: Command @unchecked ⇒
+        // the above type-check does nothing, since Command is tun
+        // we cast explicitly to fail early in case of type mismatch
+        val c = command.asInstanceOf[Command]
+        phase.onCommand(c)
+    }
+  }
+
+  // ----------
+
+  def applySideEffects(effects: immutable.Seq[ChainableEffect[_, S]]): Behavior[Any] = {
+    var res: Behavior[Any] = same
+    val it = effects.iterator
+
+    // if at least one effect results in a `stop`, we need to stop
+    // manual loop implementation to avoid allocations and multiple scans
+    while (it.hasNext) {
+      val effect = it.next()
+      applySideEffect(effect) match {
+        case _: StoppedBehavior[_] ⇒ res = Behaviors.stopped
+        case _                     ⇒ // nothing to do
+      }
+    }
+
+    res
+  }
+
+  def applySideEffect(effect: ChainableEffect[_, S]): Behavior[Any] = effect match {
+    case _: Stop.type @unchecked ⇒
+      Behaviors.stopped
+
+    case SideEffect(callbacks) ⇒
+      callbacks(state)
+      same
+  }
+
+  def applyEvent(s: S, event: E): S =
+    eventHandler(s, event)
+
+  @tailrec private def applyEffects(msg: Any, effect: Effect[E, S], sideEffects: immutable.Seq[ChainableEffect[_, S]] = Nil): Behavior[Any] = {
+    log.info(s"ApplyEffects: ${msg} => Effect: ${effect}")
+
+    effect match {
+      case CompositeEffect(Some(persist), currentSideEffects) ⇒
+        applyEffects(msg, persist, currentSideEffects ++ sideEffects) // "unwrap" & recur
+
+      case CompositeEffect(_, currentSideEffects) ⇒
+        applySideEffects(currentSideEffects ++ sideEffects) unlessStopped tryUnstash(context, this)
+
+      case Persist(event) ⇒
+        // apply the event before persist so that validation exception is handled before persisting
+        // the invalid event, in case such validation is implemented in the event handler.
+        // also, ensure that there is an event handler for each single event
+        state = applyEvent(state, event)
+        val tags = tagger(event)
+        val eventToPersist = if (tags.isEmpty) event else Tagged(event, tags)
+
+        log.info(s" >>>> [${context.self.path.name}] in Persist($event) @ ${lastSequenceNr}")
+
+        internalPersist(eventToPersist) { _ ⇒
+          applySideEffects(sideEffects)
```

Comment:
the returned behavior is not used (stopping?)

Reply:
Good catch, and a bug -- working on it

Reply:
This one is a bit tricky... will figure it out